system/lib/wasmfs/js_file_backend.cpp

```diff
@@ -0,0 +1,82 @@
+// Copyright 2021 The Emscripten Authors.  All rights reserved.
+// Emscripten is available under two separate licenses, the MIT license and the
+// University of Illinois/NCSA Open Source License.  Both these licenses can be
+// found in the LICENSE file.
+
+// This file defines the JS file backend and JS file of the new file system.
+// Current Status: Work in Progress.
+// See https://github.com/emscripten-core/emscripten/issues/15041.
+
+#include "backend.h"
+#include "wasmfs.h"
+
+using js_index_t = uint32_t;
+
+extern "C" {
+void _emscripten_write_js_file(js_index_t index,
+                               const uint8_t* buffer,
+                               size_t length,
+                               off_t offset);
+void _emscripten_read_js_file(js_index_t index,
+                              const uint8_t* buffer,
+                              size_t length,
+                              off_t offset);
+int _emscripten_get_js_file_size(js_index_t index);
+int _emscripten_create_js_file();
+void _emscripten_remove_js_file(js_index_t index);
+}
+
+namespace wasmfs {
+
+// This class describes a file that lives in JS Memory
+class JSFile : public DataFile {
+  // This index indicates the location of the JS File in the backing JS array.
+  js_index_t index;
+
+  // JSFiles will write from a Wasm Memory buffer into the backing JS array.
+  __wasi_errno_t write(const uint8_t* buf, size_t len, off_t offset) override {
+
+    _emscripten_write_js_file(index, buf, len, offset);
+
+    return __WASI_ERRNO_SUCCESS;
+  }
+
+  // JSFiles will read from the backing JS array into a Wasm Memory buffer.
+  __wasi_errno_t read(uint8_t* buf, size_t len, off_t offset) override {
+    assert(offset + len - 1 < getSize());
+
+    _emscripten_read_js_file(index, buf, len, offset);
+
+    return __WASI_ERRNO_SUCCESS;
+  }
+
+  // The size of the JSFile is defined as the length of the backing JS array.
+  size_t getSize() override { return _emscripten_get_js_file_size(index); }
+
+public:
+  JSFile(mode_t mode, backend_t backend) : DataFile(mode, backend) {
+    // Create a new file in the backing JS array and store its index.
+    index = _emscripten_create_js_file();
+  }
+
+  // Remove the typed array file contents in the backing JS array.
+  ~JSFile() { _emscripten_remove_js_file(index); }
```

Comment:
```suggestion
  ~JSFile() override { _emscripten_remove_js_file(index); }
```

Why are `write`, `read`, and `getSize` private? I would have expected them to be public.

Reply:
They are private to the file itself. We want all operations to be accessed via the `Handle` functions since we want caller functions to obtain a locked `JSFile`.